### PR TITLE
WIP: stdin support for FASTA/FASTQ files

### PR DESCRIPTION
**Pull request by RaeedA**
Issue #2017 
Adding support for FASTA/FASTQ files to make them accept stdin streaming, as well as adding compatibility for other file formats in the future.
Modified the FASTQ file parser to match the FASTA structure, specifically moving the parser functionality out of the reader and into a dedicated helper function.
Making most sniffer/parsers pass out consumed lines.
Coverage still in progress.

Please complete the following checklist:

* [x] I have read the [contribution guidelines](https://scikit.bio/contribute.html).

* [x] I have documented all public-facing changes in the [changelog](https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md).

* [ ] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/scikit-bio/scikit-bio/blob/main/LICENSE.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied.

  - **It is your responsibility to disclose** code, documentation, or other content derived from external source(s). If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [x] This pull request does not include code, documentation, or other content derived from external source(s).

Note: [This document](https://scikit.bio/devdoc/review.html) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

### CODE REVIEW

File: skbio/io/_iosources.py [hunk 1]
```diff
@@ -190,6 +201,9 @@ def can_read(self):
             if isinstance(head, str):
```
Comment (qiyunzhu) [inline skbio/io/_iosources.py:201]: These two checks can be merged:

```python
if isinstance(head, (str, bytes)):
    self.repaired = itertools.chain([head], iterator)
    return True
```

File: skbio/io/format/_base.py [hunk 2]
```diff
@@ -227,13 +229,18 @@ def _line_generator(fh, skip_blanks=False, strip=True):
 def _too_many_blanks(fh, max_blanks):
     count = 0
     too_many = False
+    consumed = []
     for line in _line_generator(fh, skip_blanks=False):
+        consumed += [line]
```
Comment (qiyunzhu) [inline skbio/io/format/_base.py:234]: This can be replaced with:

```python
consumed.append(line)
```

File: skbio/io/format/fasta.py [hunk 3]
```diff
@@ -640,22 +640,26 @@ def _fasta_sniffer(fh):
     #   not be validated but it probably isn't what the user wanted). Also, if
     #   we add QUAL as its own file format in the future, we wouldn't want the
     #   FASTA and QUAL sniffers to both positively identify a QUAL file.
-    if _too_many_blanks(fh, 5):
-        return False, {}
+    blanks, consumed = _too_many_blanks(fh, 5)
+    if blanks:
+        return False, {}, consumed
+    if not fh.seekable():
+        fh = itertools.chain(consumed, fh)
```
Comment (qiyunzhu) [inline skbio/io/format/fasta.py:647]: Very good. It looks like the right solution.

File: skbio/io/format/_base.py (hunk 2 above)
Comment (RaeedA) [reply skbio/io/format/_base.py:234]: Is this more efficient than just doing +=? I'm using += in other places as sometimes I need to add multiple lines, and wanted to keep it consistent.

File: skbio/io/registry.py [hunk 4]
```diff
@@ -514,24 +519,37 @@ def _read_gen(self, file, fmt, into, verify, kwargs):
...
+                try:
+                    # Attempt to reset and sniff file
+                    backup = file.tell()
+                    res = sniffer(file, **io_kwargs)
+                    is_format = res[0]
+                    skwargs = res[1]
+                    file.seek(backup)
+                except io.UnsupportedOperation:
+                    # File is not seekable, so get consumed lines to reset it
+                    is_format, skwargs, consumed = sniffer(file, **io_kwargs)
```
Comment (wasade) [inline skbio/io/registry.py:551]: `peek` for `stdin` (see `sys.stdin.buffer.peek`) and many other IO types (not `io.StringIO` AFAIK) would work without advancing the pointer too. But, it's read request is not assured to be what you want in terms of bytes

File: skbio/io/registry.py [hunk 5]
```diff
@@ -914,7 +939,8 @@ def wrapped_sniffer(
                         # Some formats may have headers which indicate their
                         # format sniffers should be able to rely on the
                         # filehandle to point at the beginning of the file.
-                        fh.seek(0)
+                        if self.header:
+                            fh.seek(0)
```
Comment (wasade) [inline skbio/io/registry.py:943]: Is this still safe for `stdin` which is not seekable?

File: skbio/io/tests/test_util.py [hunk 6]
```diff
@@ -634,6 +638,27 @@ def test_open_write_mode(self):
         fh.close()
         self.assertTrue(fh.closed)
 
+class TestReadStandardInput(unittest.TestCase):
+    num_text = ["ONE", "TWO", "THREE", "FOUR", "FIVE", "SIX", "SEVEN"]
+
+    def test_stdin_read_fasta(self):
```
Comment (wasade) [inline skbio/io/tests/test_util.py:644]: These tests should stay, but I recommend additionally setting up an independent integration test run through the `Makefile`. Something like the below. The inline python though could also be placed in a minor script. 

```bash
#!/bin/bash

cat skbio/io/tests/test_data/some_fasta.fna | \
    python -c "import skbio, sys; print([r.metadata['id'] for r in skbio.read(sys.stdin, format='fasta')]) > /tmp/obs_stdin.ids

grep "^>" skbio/io/tests/test_data/some_fasta.fna | tr -d ">" > /tmp/exp_stdin.ids

obs_md5=$(md5sum /tmp/obs_stdin.ids)
exp_md5=$(md5sum /tmp/exp_stdin.ids)

if [[ ${obs_md5} != ${exp_md5} ]]; then
    exit 1
fi
```

File: skbio/io/format/_base.py (hunk 2 above)
Comment (qiyunzhu) [reply skbio/io/format/_base.py:234]: `append` is just a "pythonic" way :) I don't know if it's more efficient or not. PS: if  you want to add multiple elements, you can do `consumed.extend([line1, line2, line3...])`.

File: skbio/io/registry.py (hunk 5 above)
Comment (qiyunzhu) [reply skbio/io/registry.py:943]: I think that @RaeedA added a `seekable` flag to different input types to control this behavior.

Comment (RaeedA) [reply skbio/io/registry.py:943]: That was the original plan, I moved some things around in the current changes to make it work more intuitively. It's now in a try/catch block.

**Comment by wasade** [comment]
Does this work?How to check if file object is random accessstackoverflow.comOn Jun 3, 2024, at 15:05, Raeed Azom ***@***.***> wrote:﻿
@RaeedA commented on this pull request.



In skbio/io/registry.py:
> @@ -914,7 +939,8 @@ def wrapped_sniffer(
                         # Some formats may have headers which indicate their
                         # format sniffers should be able to rely on the
                         # filehandle to point at the beginning of the file.
-                        fh.seek(0)
+                        if self.header:
+                            fh.seek(0)

That was the original plan, I moved some things around in the current changes to make it work more intuitively. It's now in a try/catch block.

**Comment by qiyunzhu** [comment]
@wasade According to the Python [documentation](https://docs.python.org/3/library/io.html#io.IOBase.seekable), `seekable` is a valid check. It says: "Return True if the stream supports random access".

**Comment by wasade** [comment]
Right, and it would be preferable to perform the explicit check rather than a try/except

**Comment by wasade** [comment]
Oh I see, the SO post didn't actually come through in email -- this [one](https://stackoverflow.com/a/49376762/19741)

**Comment by qiyunzhu** [comment]
> Right, and it would be preferable to perform the explicit check rather than a try/except

I agree with that. In the current PR there are both. Redundancy may not be bad. But @RaeedA you may want to check to make sure we don't solely rely on try/except. (I am okay with solely relying on seekable, if that works.)

**Comment by RaeedA** [comment]
Hey! I'm actually not using the seekable call here, as I've found that it behaves irregularly with some data formats. I've instead been using tell() wrapped inside of a try/catch, as tell will consistently produce an UnsupportedOperation error that I can use to detect seekable versus unseekable.

File: skbio/io/tests/test_util.py (hunk 6 above)
Comment (RaeedA) [reply skbio/io/tests/test_util.py:644]: I'm not super familiar with these, but I've been looking for a way to do just this! I'll take a look at implementing this for testing.

File: skbio/io/registry.py (hunk 4 above)
Comment (RaeedA) [reply skbio/io/registry.py:551]: Yeah, I've tested a couple other methods of reading nonseekables including this but it just seems that reconstructing it as a chain is the best.

**Comment by qiyunzhu** [comment]
@RaeedA I tend to think that using try/except is fine, given that you found its result consistent, whereas that of seekable isn't.

**Pull request closed without merging**